From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stephen Chenney <schenney@chromium.org>
Date: Tue, 18 Aug 2026 16:21:41 -0700
Subject: [HiC] Privacy protection for pseudo elements

Protect private content in pseudo elements by setting IsInCanvasSubtree
for all pseudo elements. Add tests to cover the following:
::first-line
::first-letter
::before and ::after
::marker
::details-content
::details-content::after (to test nested pseudos)

There are still more to test but land this first and then add more,
to get the protection in sooner. Testing everything will require
several more days.

Fixed: 532904047
Change-Id: I859b78904b7ac8980c548db593274c009c16fdb2
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/8253249
Reviewed-by: Stefan Zager <szager@chromium.org>
Commit-Queue: Stephen Chenney <schenney@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1681817}

diff --git a/third_party/blink/renderer/core/dom/element.cc b/third_party/blink/renderer/core/dom/element.cc
index 97bbb2a8bd913ed3b482741390f5726d667c4208..9defc3828f9cc014dee957afba71d21f1a4402a4 100644
--- a/third_party/blink/renderer/core/dom/element.cc
+++ b/third_party/blink/renderer/core/dom/element.cc
@@ -4399,6 +4399,11 @@ void Element::SetIsCanvasOrInCanvasSubtree(bool value) {
     return;
   }
   SetElementFlag(ElementFlags::kIsCanvasOrInCanvasSubtree, value);
+  if (const ElementRareDataVector* rare_data = RareData()) {
+    for (PseudoElement* pseudo_element : rare_data->GetPseudoElements()) {
+      pseudo_element->SetIsCanvasOrInCanvasSubtree(value);
+    }
+  }
   DidChangeIsCanvasOrInCanvasSubtree();
 }
 
diff --git a/third_party/blink/renderer/core/paint/object_paint_invalidator.cc b/third_party/blink/renderer/core/paint/object_paint_invalidator.cc
index 57cbac0cc543d0bce6a843142b159e02cbe239c8..f3d2535780acd01869934749cf1dac05f84dd4a4 100644
--- a/third_party/blink/renderer/core/paint/object_paint_invalidator.cc
+++ b/third_party/blink/renderer/core/paint/object_paint_invalidator.cc
@@ -34,7 +34,7 @@ void ObjectPaintInvalidator::InvalidateDisplayItemClient(
   // It's caller's responsibility to ensure PaintingLayer's NeedsRepaint is
   // set. Don't set the flag here because getting PaintLayer has cost and the
   // caller can use various ways (e.g.
-  // PaintInvalidatinContext::painting_layer) to reduce the cost.
+  // PaintInvalidationContext::painting_layer) to reduce the cost.
   CheckPaintLayerNeedsRepaint();
 #endif
   TRACE_EVENT_INSTANT2(TRACE_DISABLED_BY_DEFAULT("blink.invalidation"),
diff --git a/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/before-after-images-ignored.tentative.https.sub.html b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/before-after-images-ignored.tentative.https.sub.html
new file mode 100644
index 0000000000000000000000000000000000000000..89b2191c79e2d57a738af004e3cbad0d6ed3100f
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/before-after-images-ignored.tentative.https.sub.html
@@ -0,0 +1,56 @@
+<!DOCTYPE HTML>
+<html>
+<head>
+  <title>drawElementImage does not draw cross-origin images for before/after</title>
+  <script src='/resources/testharness.js'></script>
+  <script src='/resources/testharnessreport.js'></script>
+  <script src='/html/canvas/resources/canvas-tests.js'></script>
+  <script src="/html/canvas/resources/wait-for-canvas-paint.js"></script>
+  <style>
+    .wrapper {
+      width: 400px;
+      height: 120px;
+    }
+    .base::before {
+      content: url("https://{{hosts[alt][www]}}:{{ports[https][0]}}/images/red-100x100.png");
+      width: 100px;
+      height: 100px;
+      background-color: green;
+      display: inline-block;
+    }
+    .base::after {
+      content: url("https://{{hosts[alt][www]}}:{{ports[https][0]}}/images/red-100x100.png");
+      width: 100px;
+      height: 100px;
+      background-color: green;
+      display: inline-block;
+    }
+  </style>
+</head>
+<body>
+  <canvas id=canvas width="100" height="200" layoutsubtree>
+    <div id=child class=wrapper>
+      <p class=base>Before/After</p>
+    </div>
+  </canvas>
+
+  <script>
+    window.onload = () => {
+      promise_test(async function(t) {
+        await waitForCanvasPaint(canvas);
+        const context = canvas.getContext("2d");
+        context.drawElementImage(child, 0, 0);
+
+        // Fetch all pixel data once to avoid multiple slow readbacks.
+        const imgData = context.getImageData(0, 0, canvas.width, canvas.height);
+
+        pixel = _getPixelFromImageData(imgData, 50, 50);
+        assert_array_equals(pixel, [0, 128, 0, 255], "Cross origin ::before should not draw");
+
+        pixel = _getPixelFromImageData(imgData, 250, 50);
+        assert_array_equals(pixel, [0, 128, 0, 255], "Cross origin ::after should not draw");
+      });
+    };
+  </script>
+</body>
+</html>
diff --git a/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/before-after-images-moved-ignored.tentative.https.sub.html b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/before-after-images-moved-ignored.tentative.https.sub.html
new file mode 100644
index 0000000000000000000000000000000000000000..0eb6f3b439c095454f80e382010db28722159c18
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/before-after-images-moved-ignored.tentative.https.sub.html
@@ -0,0 +1,60 @@
+<!DOCTYPE HTML>
+<html>
+<head>
+  <title>drawElementImage does not draw cross-origin images for before/after</title>
+  <script src='/resources/testharness.js'></script>
+  <script src='/resources/testharnessreport.js'></script>
+  <script src='/html/canvas/resources/canvas-tests.js'></script>
+  <script src="/html/canvas/resources/wait-for-canvas-paint.js"></script>
+  <style>
+    .wrapper {
+      width: 400px;
+      height: 120px;
+    }
+    .base::before {
+      content: url("https://{{hosts[alt][www]}}:{{ports[https][0]}}/images/red-100x100.png");
+      width: 100px;
+      height: 100px;
+      background-color: green;
+      display: inline-block;
+    }
+    .base::after {
+      content: url("https://{{hosts[alt][www]}}:{{ports[https][0]}}/images/red-100x100.png");
+      width: 100px;
+      height: 100px;
+      background-color: green;
+      display: inline-block;
+    }
+  </style>
+</head>
+<body>
+  <canvas id=canvas width="100" height="200" layoutsubtree>
+    <div id=child class=wrapper>
+    </div>
+  </canvas>
+  <p id="text" class=base>Before/After</p>
+
+  <script>
+    window.onload = () => {
+      promise_test(async function(t) {
+        await waitForCanvasPaint(canvas);
+
+        child.moveBefore(text, null);
+
+        await waitForCanvasPaint(canvas);
+        const context = canvas.getContext("2d");
+        context.drawElementImage(child, 0, 0);
+
+        // Fetch all pixel data once to avoid multiple slow readbacks.
+        const imgData = context.getImageData(0, 0, canvas.width, canvas.height);
+
+        pixel = _getPixelFromImageData(imgData, 50, 50);
+        assert_array_equals(pixel, [0, 128, 0, 255], "Cross origin ::before should not draw");
+
+        pixel = _getPixelFromImageData(imgData, 250, 50);
+        assert_array_equals(pixel, [0, 128, 0, 255], "Cross origin ::after should not draw");
+      });
+    };
+  </script>
+</body>
+</html>
diff --git a/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/before-after-visited-color-ignored.tentative.html b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/before-after-visited-color-ignored.tentative.html
new file mode 100644
index 0000000000000000000000000000000000000000..977142ce2b22a8aa33209accd50ce95dbdc327e4
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/before-after-visited-color-ignored.tentative.html
@@ -0,0 +1,64 @@
+<!DOCTYPE HTML>
+<html>
+<head>
+  <title>drawElementImage does not draw visited colors for before/after</title>
+  <script src='/resources/testharness.js'></script>
+  <script src='/resources/testharnessreport.js'></script>
+  <script src='/html/canvas/resources/canvas-tests.js'></script>
+  <script src="/html/canvas/resources/wait-for-canvas-paint.js"></script>
+  <style>
+    a {
+      color: green;
+    }
+    a:visited {
+      color: red;
+    }
+    .wrapper {
+      width: 400px;
+      height: 120px;
+    }
+    .base::before {
+      content: "";
+      width: 100px;
+      height: 100px;
+      background-color: currentColor;
+      display: inline-block;
+    }
+    .base::after {
+      content: "";
+      width: 100px;
+      height: 100px;
+      background-color: currentColor;
+      display: inline-block;
+    }
+  </style>
+</head>
+<body>
+  <a href="">
+    <canvas id=canvas width="100" height="200" layoutsubtree>
+      <div id=child class=wrapper>
+        <p class=base>Before/After</p>
+      </div>
+    </canvas>
+  </a>
+
+  <script>
+    window.onload = () => {
+      promise_test(async function(t) {
+        await waitForCanvasPaint(canvas);
+        const context = canvas.getContext("2d");
+        context.drawElementImage(child, 0, 0);
+
+        // Fetch all pixel data once to avoid multiple slow readbacks.
+        const imgData = context.getImageData(0, 0, canvas.width, canvas.height);
+
+        pixel = _getPixelFromImageData(imgData, 50, 50);
+        assert_array_equals(pixel, [0, 128, 0, 255], "Visited color in ::before should not draw");
+
+        pixel = _getPixelFromImageData(imgData, 250, 50);
+        assert_array_equals(pixel, [0, 128, 0, 255], "Visited color in ::after should not draw");
+      });
+    };
+  </script>
+</body>
+</html>
diff --git a/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/before-background-image-iframe-ignored.tentative.https.sub.html b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/before-background-image-iframe-ignored.tentative.https.sub.html
new file mode 100644
index 0000000000000000000000000000000000000000..2f236942a0e9efb426cfcf1a96d41bdcdb5d6718
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/before-background-image-iframe-ignored.tentative.https.sub.html
@@ -0,0 +1,39 @@
+<!DOCTYPE HTML>
+<html>
+<head>
+  <title>drawElementImage does not draw cross-origin background images for before/after</title>
+  <script src='/resources/testharness.js'></script>
+  <script src='/resources/testharnessreport.js'></script>
+  <script src='/html/canvas/resources/canvas-tests.js'></script>
+  <script src="/html/canvas/resources/wait-for-canvas-paint.js"></script>
+  <style>
+    iframe {
+      width: 100px;
+      height: 200px;
+      background-color: blue;
+    }
+  </style>
+</head>
+<body>
+  <canvas id=canvas width="100" height="200" layoutsubtree>
+    <div id=child>
+      <iframe id=iframe src="../support/subframe-cross-origin-before-background-on-body.https.sub.html"></iframe>
+    </div>
+  </canvas>
+
+  <script>
+    iframe.onload = () => {
+      promise_test(async function(t) {
+        await waitForCanvasPaint(canvas);
+        const context = canvas.getContext("2d");
+        context.drawElementImage(child, 0, 0);
+
+        let pixel = context.getImageData(20, 20, 1, 1).data;
+        assert_false(pixel[0] > 0, "The canvas should not paint cross origin content.");
+        assert_true(pixel[1] > 0, "The canvas should not paint cross origin content.");
+        assert_false(pixel[2] > 0, "The canvas should not paint cross origin content.");
+      });
+    };
+  </script>
+</body>
+</html>
diff --git a/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/before-background-image-iframe-modified-ignored.tentative.https.sub.html b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/before-background-image-iframe-modified-ignored.tentative.https.sub.html
new file mode 100644
index 0000000000000000000000000000000000000000..4f709843665d56eb19a3c5a09e1ea4ae29382d8b
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/before-background-image-iframe-modified-ignored.tentative.https.sub.html
@@ -0,0 +1,50 @@
+<!DOCTYPE HTML>
+<html>
+<head>
+  <title>drawElementImage does not draw cross-origin background images for before/after when modified</title>
+  <script src='/resources/testharness.js'></script>
+  <script src='/resources/testharnessreport.js'></script>
+  <script src='/html/canvas/resources/canvas-tests.js'></script>
+  <script src="/html/canvas/resources/wait-for-canvas-paint.js"></script>
+  <style>
+    iframe {
+      width: 100px;
+      height: 200px;
+      background-color: blue;
+    }
+  </style>
+</head>
+<body>
+  <canvas id=canvas width="100" height="200" layoutsubtree>
+    <div id=child>
+      <iframe id=iframe src="../support/subframe-cross-origin-before-background-on-body-modified.https.sub.html"></iframe>
+    </div>
+  </canvas>
+
+  <script>
+    iframe.onload = () => {
+      promise_test(async function(t) {
+        await waitForCanvasPaint(canvas);
+        const context = canvas.getContext("2d");
+        context.drawElementImage(child, 0, 0);
+
+        let pixel = context.getImageData(20, 20, 1, 1).data;
+        assert_false(pixel[0] > 0, "Initial draw: Red should not paint.");
+        assert_false(pixel[1] > 0, "Initial draw: Green should not paint before class is added.");
+        assert_true(pixel[2] > 0, "Initial draw: Blue iframe background should paint.");
+
+        // Add class to iframe body causing it to match the ::before rule.
+        iframe.contentDocument.body.classList.add('has-before');
+
+        await waitForCanvasPaint(canvas);
+        context.drawElementImage(child, 0, 0);
+
+        pixel = context.getImageData(20, 20, 1, 1).data;
+        assert_false(pixel[0] > 0, "The canvas should not paint cross origin content.");
+        assert_true(pixel[1] > 0, "The canvas should paint the ::before background color.");
+        assert_false(pixel[2] > 0, "The canvas should paint the ::before element over the iframe background.");
+      });
+    };
+  </script>
+</body>
+</html>
diff --git a/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/before-background-image-iframe-moved-ignored.tentative.https.sub.html b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/before-background-image-iframe-moved-ignored.tentative.https.sub.html
new file mode 100644
index 0000000000000000000000000000000000000000..40b84e66ca245e6ee7cc512fcea3e35a7b00b99f
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/before-background-image-iframe-moved-ignored.tentative.https.sub.html
@@ -0,0 +1,44 @@
+<!DOCTYPE HTML>
+<html>
+<head>
+  <title>drawElementImage does not draw cross-origin background images for before/after</title>
+  <script src='/resources/testharness.js'></script>
+  <script src='/resources/testharnessreport.js'></script>
+  <script src='/html/canvas/resources/canvas-tests.js'></script>
+  <script src="/html/canvas/resources/wait-for-canvas-paint.js"></script>
+  <style>
+    iframe {
+      width: 100px;
+      height: 200px;
+      background-color: blue;
+    }
+  </style>
+</head>
+<body>
+  <canvas id=canvas width="100" height="200" layoutsubtree>
+    <div id=child>
+    </div>
+  </canvas>
+  <iframe id=iframe src="../support/subframe-cross-origin-before-background-on-body.https.sub.html"></iframe>
+
+  <script>
+    iframe.onload = () => {
+      promise_test(async function(t) {
+        await waitForCanvasPaint(canvas);
+        const context = canvas.getContext("2d");
+
+        // Atomically move the iframe into the canvas.
+        child.moveBefore(iframe, null);
+
+        await waitForCanvasPaint(canvas);
+        context.drawElementImage(child, 0, 0);
+
+        let pixel = context.getImageData(20, 20, 1, 1).data;
+        assert_false(pixel[0] > 0, "The canvas should not paint cross origin content.");
+        assert_true(pixel[1] > 0, "The canvas should not paint cross origin content.");
+        assert_false(pixel[2] > 0, "The canvas should not paint cross origin content.");
+      });
+    };
+  </script>
+</body>
+</html>
diff --git a/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/before-images-iframe-ignored.tentative.https.sub.html b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/before-images-iframe-ignored.tentative.https.sub.html
new file mode 100644
index 0000000000000000000000000000000000000000..ebef3136a2b12063c4c738137f0961c48d62a005
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/before-images-iframe-ignored.tentative.https.sub.html
@@ -0,0 +1,39 @@
+<!DOCTYPE HTML>
+<html>
+<head>
+  <title>drawElementImage does not draw cross-origin images for before/after</title>
+  <script src='/resources/testharness.js'></script>
+  <script src='/resources/testharnessreport.js'></script>
+  <script src='/html/canvas/resources/canvas-tests.js'></script>
+  <script src="/html/canvas/resources/wait-for-canvas-paint.js"></script>
+  <style>
+    iframe {
+      width: 100px;
+      height: 200px;
+      background-color: blue;
+    }
+  </style>
+</head>
+<body>
+  <canvas id=canvas width="100" height="200" layoutsubtree>
+    <div id=child>
+      <iframe id=iframe src="../support/subframe-cross-origin-before-content-on-body.https.sub.html"></iframe>
+    </div>
+  </canvas>
+
+  <script>
+    iframe.onload = () => {
+      promise_test(async function(t) {
+        await waitForCanvasPaint(canvas);
+        const context = canvas.getContext("2d");
+        context.drawElementImage(child, 0, 0);
+
+        let pixel = context.getImageData(20, 20, 1, 1).data;
+        assert_false(pixel[0] > 0, "The canvas should not paint cross origin content.");
+        assert_true(pixel[1] > 0, "The canvas should not paint cross origin content.");
+        assert_false(pixel[2] > 0, "The canvas should not paint cross origin content.");
+      });
+    };
+  </script>
+</body>
+</html>
diff --git a/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/before-images-iframe-moved-ignored.tentative.https.sub.html b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/before-images-iframe-moved-ignored.tentative.https.sub.html
new file mode 100644
index 0000000000000000000000000000000000000000..2ac4a6cfec1733bf0f7a2c14969ae6d68e39cb8e
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/before-images-iframe-moved-ignored.tentative.https.sub.html
@@ -0,0 +1,44 @@
+<!DOCTYPE HTML>
+<html>
+<head>
+  <title>drawElementImage does not draw cross-origin images for before/after</title>
+  <script src='/resources/testharness.js'></script>
+  <script src='/resources/testharnessreport.js'></script>
+  <script src='/html/canvas/resources/canvas-tests.js'></script>
+  <script src="/html/canvas/resources/wait-for-canvas-paint.js"></script>
+  <style>
+    iframe {
+      width: 100px;
+      height: 200px;
+      background-color: blue;
+    }
+  </style>
+</head>
+<body>
+  <canvas id=canvas width="100" height="200" layoutsubtree>
+    <div id=child>
+    </div>
+  </canvas>
+  <iframe id=iframe src="../support/subframe-cross-origin-before-content-on-body.https.sub.html"></iframe>
+
+  <script>
+    iframe.onload = () => {
+      promise_test(async function(t) {
+        await waitForCanvasPaint(canvas);
+        const context = canvas.getContext("2d");
+
+        // Atomically move the iframe into the canvas.
+        child.moveBefore(iframe, null);
+
+        await waitForCanvasPaint(canvas);
+        context.drawElementImage(child, 0, 0);
+
+        let pixel = context.getImageData(20, 20, 1, 1).data;
+        assert_false(pixel[0] > 0, "The canvas should not paint cross origin content.");
+        assert_true(pixel[1] > 0, "The canvas should not paint cross origin content.");
+        assert_false(pixel[2] > 0, "The canvas should not paint cross origin content.");
+      });
+    };
+  </script>
+</body>
+</html>
diff --git a/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/details-content-after-image-ignored.tentative.https.sub.html b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/details-content-after-image-ignored.tentative.https.sub.html
new file mode 100644
index 0000000000000000000000000000000000000000..4400a1a1d42edaf474a17192d24d72314eeab48d
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/details-content-after-image-ignored.tentative.https.sub.html
@@ -0,0 +1,53 @@
+<!DOCTYPE HTML>
+<html>
+<head>
+  <title>drawElementImage does not paint corss-origin ::after image in ::details-content</title>
+  <link rel="stylesheet" type="text/css" href="/fonts/ahem.css" />
+  <script src='/resources/testharness.js'></script>
+  <script src='/resources/testharnessreport.js'></script>
+  <script src='/html/canvas/resources/canvas-tests.js'></script>
+  <script src="/html/canvas/resources/wait-for-canvas-paint.js"></script>
+  <style>
+    details {
+      width: 100px;
+    }
+    ::details-content::after {
+      content: url("https://{{hosts[alt][www]}}:{{ports[https][0]}}/images/red-100x100.png");
+      width: 100px;
+      height: 100px;
+      background-color: green;
+      display: inline-block;
+    }
+  </style>
+</head>
+<body>
+  <canvas id="canvas" width="100" height="200" layoutsubtree>
+    <div id=child>
+      <details open>
+        <summary>Details</summary>
+        <p>A</p>
+      </details>
+    </div>
+  </canvas>
+
+  <script>
+    window.onload = () => {
+      promise_test(async function(t) {
+        await waitForCanvasPaint(canvas);
+
+        var context = canvas.getContext("2d");
+
+        // Draw the element image
+        context.drawElementImage(child, 0, 0);
+
+        const imgData = context.getImageData(0, 0, canvas.width, canvas.height);
+
+        // The canvas should NOT paint the background image (red)
+        pixel = _getPixelFromImageData(imgData, 50, 80);
+        assert_false(pixel[0] > 0, "The canvas should not paint a cross origin image.");
+        assert_true(pixel[1] > 0, "The canvas should show the background color.");
+      });
+    };
+  </script>
+</body>
+</html>
diff --git a/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/details-content-background-image-ignored.tentative.https.sub.html b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/details-content-background-image-ignored.tentative.https.sub.html
new file mode 100644
index 0000000000000000000000000000000000000000..2ff7dbc05bc12107aa9d0683a0abd81126b77886
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/details-content-background-image-ignored.tentative.https.sub.html
@@ -0,0 +1,51 @@
+<!DOCTYPE HTML>
+<html>
+<head>
+  <title>drawElementImage does not paint corss-origin background in ::details-content</title>
+  <link rel="stylesheet" type="text/css" href="/fonts/ahem.css" />
+  <script src='/resources/testharness.js'></script>
+  <script src='/resources/testharnessreport.js'></script>
+  <script src='/html/canvas/resources/canvas-tests.js'></script>
+  <script src="/html/canvas/resources/wait-for-canvas-paint.js"></script>
+  <style>
+    details {
+      width: 100px;
+    }
+    ::details-content {
+      color: transparent;
+      background-color: green;
+      background-image: url("https://{{hosts[alt][www]}}:{{ports[https][0]}}/images/red-100x100.png");
+    }
+  </style>
+</head>
+<body>
+  <canvas id="canvas" width="100" height="100" layoutsubtree>
+    <div id=child>
+      <details open>
+        <summary>Details</summary>
+        <a class=link href="">A</a>
+      </details>
+    </div>
+  </canvas>
+
+  <script>
+    window.onload = () => {
+      promise_test(async function(t) {
+        await waitForCanvasPaint(canvas);
+
+        var context = canvas.getContext("2d");
+
+        // Draw the element image
+        context.drawElementImage(child, 0, 0);
+
+        const imgData = context.getImageData(0, 0, canvas.width, canvas.height);
+
+        // The canvas should NOT paint the background image (red)
+        pixel = _getPixelFromImageData(imgData, 10, 25);
+        assert_false(pixel[0] > 0, "The canvas should not paint a cross origin image.");
+        assert_true(pixel[1] > 0, "The canvas should show the background color.");
+      });
+    };
+  </script>
+</body>
+</html>
diff --git a/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/details-content-visited-link-currentcolor-ignored.tentative.html b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/details-content-visited-link-currentcolor-ignored.tentative.html
new file mode 100644
index 0000000000000000000000000000000000000000..fa80653dfb59bf24caed4265ff6f9a1887c9e25b
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/details-content-visited-link-currentcolor-ignored.tentative.html
@@ -0,0 +1,70 @@
+<!DOCTYPE HTML>
+<html>
+<head>
+  <title>drawElementImage does not reveal visited link colors in ::details-content</title>
+  <link rel="stylesheet" type="text/css" href="/fonts/ahem.css" />
+  <script src='/resources/testharness.js'></script>
+  <script src='/resources/testharnessreport.js'></script>
+  <script src='/html/canvas/resources/canvas-tests.js'></script>
+  <script src="/html/canvas/resources/wait-for-canvas-paint.js"></script>
+  <style>
+    .link {
+      color: green;
+    }
+    .link:visited {
+      color: red;
+    }
+    details {
+      width: 100px;
+    }
+    #details1::details-content {
+      font-family: Ahem;
+      color: currentColor;
+      font-size: 20px;
+    }
+    #details2::details-content {
+      font-family: Ahem;
+      background-color: currentColor;
+      font-size: 20px;
+    }
+  </style>
+</head>
+<body>
+  <canvas id="canvas" width="100" height="100" layoutsubtree>
+    <a id=child class=link href="">
+      <details id=details1 open>
+        <summary>Details</summary>
+        <p>A</p>
+      </details>
+      <details id=details2 open>
+        <summary>Details</summary>
+        <p>-</p>
+      </details>
+    </a>
+  </canvas>
+
+  <script>
+    window.onload = () => {
+      promise_test(async function(t) {
+        await waitForCanvasPaint(canvas);
+
+        var context = canvas.getContext("2d");
+
+        // Draw the element image
+        context.drawElementImage(child, 0, 0);
+
+        const imgData = context.getImageData(0, 0, canvas.width, canvas.height);
+
+        // The canvas should NOT reflect the visited state (red), it should paint the unvisited state (green).
+        pixel = _getPixelFromImageData(imgData, 6, 45);
+        assert_false(pixel[0] > 0, "The canvas should not reflect the visited state.");
+        assert_true(pixel[1] > 0, "The canvas should not reflect the visited state.");
+
+        pixel = _getPixelFromImageData(imgData, 6, 85);
+        assert_false(pixel[0] > 0, "The canvas should not reflect the visited state in background.");
+        assert_true(pixel[1] > 0, "The canvas should not reflect the visited state in background.");
+      });
+    };
+  </script>
+</body>
+</html>
diff --git a/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/details-content-visited-link-ignored.tentative.html b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/details-content-visited-link-ignored.tentative.html
new file mode 100644
index 0000000000000000000000000000000000000000..506412863d6bc8ac6d1a27665729d1c052244490
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/details-content-visited-link-ignored.tentative.html
@@ -0,0 +1,57 @@
+<!DOCTYPE HTML>
+<html>
+<head>
+  <title>drawElementImage does not reveal visited link colors in ::details-content</title>
+  <link rel="stylesheet" type="text/css" href="/fonts/ahem.css" />
+  <script src='/resources/testharness.js'></script>
+  <script src='/resources/testharnessreport.js'></script>
+  <script src='/html/canvas/resources/canvas-tests.js'></script>
+  <script src="/html/canvas/resources/wait-for-canvas-paint.js"></script>
+  <style>
+    .link {
+      color: green;
+    }
+    .link:visited {
+      color: red;
+    }
+    details {
+      width: 100px;
+    }
+    ::details-content {
+      font-family: Ahem;
+      font-size: 20px;
+    }
+  </style>
+</head>
+<body>
+  <canvas id="canvas" width="100" height="100" layoutsubtree>
+    <div id=child>
+      <details open>
+        <summary>Details</summary>
+        <a class=link href="">A</a>
+      </details>
+    </div>
+  </canvas>
+
+  <script>
+    window.onload = () => {
+      promise_test(async function(t) {
+        await waitForCanvasPaint(canvas);
+
+        var context = canvas.getContext("2d");
+
+        // Draw the element image
+        context.drawElementImage(child, 0, 0);
+
+        const imgData = context.getImageData(0, 0, canvas.width, canvas.height);
+
+
+        // The canvas should NOT reflect the visited state (red), it should paint the unvisited state (green).
+        pixel = _getPixelFromImageData(imgData, 6, 25);
+        assert_false(pixel[0] > 0, "The canvas should not reflect the visited state.");
+        assert_true(pixel[1] > 0, "The canvas should not reflect the visited state.");
+      });
+    };
+  </script>
+</body>
+</html>
diff --git a/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/first-letter-background-images-ignored.tentative.https.sub.html b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/first-letter-background-images-ignored.tentative.https.sub.html
new file mode 100644
index 0000000000000000000000000000000000000000..d6982311b9398b8ef1be38061ad781149c5187d3
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/first-letter-background-images-ignored.tentative.https.sub.html
@@ -0,0 +1,41 @@
+<!DOCTYPE HTML>
+<html>
+<head>
+  <title>drawElementImage does not draw cross-origin background images for first-letter</title>
+  <script src='/resources/testharness.js'></script>
+  <script src='/resources/testharnessreport.js'></script>
+  <script src='/html/canvas/resources/canvas-tests.js'></script>
+  <script src="/html/canvas/resources/wait-for-canvas-paint.js"></script>
+  <style>
+    *::first-letter {
+      color: transparent;
+      background-color: green;
+      background-image: url("https://{{hosts[alt][www]}}:{{ports[https][0]}}/images/red-100x100.png");
+
+    }
+  </style>
+</head>
+<body>
+  <canvas id=canvas width="100" height="200" layoutsubtree>
+    <div id=child>
+      One line of text beginning in a big letter O.
+    </div>
+  </canvas>
+
+  <script>
+    window.onload = () => {
+      promise_test(async function(t) {
+        await waitForCanvasPaint(canvas);
+        const context = canvas.getContext("2d");
+        context.drawElementImage(child, 0, 0);
+
+        // Fetch all pixel data once to avoid multiple slow readbacks.
+        const imgData = context.getImageData(0, 0, canvas.width, canvas.height);
+
+        pixel = _getPixelFromImageData(imgData, 5, 5);
+        assert_array_equals(pixel, [0, 128, 0, 255], "Cross origin should not draw");
+      });
+    };
+  </script>
+</body>
+</html>
diff --git a/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/first-letter-visited-link-color-ignored.tentative.html b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/first-letter-visited-link-color-ignored.tentative.html
new file mode 100644
index 0000000000000000000000000000000000000000..b62f58f7e53c5478b06e86e0dea8851558db35a6
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/first-letter-visited-link-color-ignored.tentative.html
@@ -0,0 +1,55 @@
+<!DOCTYPE HTML>
+<html>
+<head>
+  <title>drawElementImage does not reveal visited link colors in first-letter</title>
+  <link rel="stylesheet" type="text/css" href="/fonts/ahem.css" />
+  <script src='/resources/testharness.js'></script>
+  <script src='/resources/testharnessreport.js'></script>
+  <script src="/html/canvas/resources/wait-for-canvas-paint.js"></script>
+  <style>
+    #link {
+      font-family: Ahem;
+      font-size: 20px;
+      background-color: green;
+      color: green;
+      display: block;
+    }
+    #link::first-letter {
+      background-color: blue;
+      color: blue;
+    }
+    #link:visited {
+      background-color: red;
+      color: red;
+    }
+    #link:visited::first-letter {
+      background-color: red;
+      color: red;
+    }
+  </style>
+</head>
+<body>
+  <canvas id="canvas" width="100" height="100" layoutsubtree>
+    <a id="link" href="">Visited color outside of canvas</a>
+  </canvas>
+
+  <script>
+    window.onload = () => {
+      promise_test(async function(t) {
+        await waitForCanvasPaint(canvas);
+
+        var context = canvas.getContext("2d");
+
+        // Draw the element image
+        context.drawElementImage(link, 0, 0);
+
+        // The canvas should NOT reflect the visited state (red), it should paint the unvisited state (green).
+        let pixel = context.getImageData(5, 5, 1, 1).data;
+        assert_false(pixel[0] > 0, "The canvas should not reflect the visited state.");
+        assert_false(pixel[1] > 0, "The canvas should not reflect the visited state.");
+        assert_true(pixel[2] > 0, "The canvas should not reflect the visited state.");
+      });
+    };
+  </script>
+</body>
+</html>
diff --git a/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/first-letter-visited-link-currentcolor-ignored.tentative.html b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/first-letter-visited-link-currentcolor-ignored.tentative.html
new file mode 100644
index 0000000000000000000000000000000000000000..b98aba1765dd90818ffbe68d51f04b81785267eb
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/first-letter-visited-link-currentcolor-ignored.tentative.html
@@ -0,0 +1,48 @@
+<!DOCTYPE HTML>
+<html>
+<head>
+  <title>drawElementImage does not reveal visited link colors first first-letter</title>
+  <link rel="stylesheet" type="text/css" href="/fonts/ahem.css" />
+  <script src='/resources/testharness.js'></script>
+  <script src='/resources/testharnessreport.js'></script>
+  <script src="/html/canvas/resources/wait-for-canvas-paint.js"></script>
+  <style>
+    p::first-letter {
+      color: currentColor;
+      background-color: currentColor;
+    }
+    .link {
+      font-family: Ahem;
+      font-size: 20px;
+      color: green;
+      display: block;
+    }
+    .link:visited {
+      color: red;
+    }
+  </style>
+</head>
+<body>
+  <canvas id="canvas" width="100" height="100" layoutsubtree>
+    <a id=child class=link href=""><p>Visited first letter.</p></a>
+  </canvas>
+
+  <script>
+    window.onload = () => {
+      promise_test(async function(t) {
+        await waitForCanvasPaint(canvas);
+
+        var context = canvas.getContext("2d");
+
+        // Draw the element image
+        context.drawElementImage(child, 0, -20);
+
+        // The canvas should NOT reflect the visited state (red), it should paint the unvisited state (green).
+        let pixel = context.getImageData(5, 5, 1, 1).data;
+        assert_false(pixel[0] > 0, "The canvas should not reflect the visited state.");
+        assert_true(pixel[1] > 0, "The canvas should not reflect the visited state.");
+      });
+    };
+  </script>
+</body>
+</html>
diff --git a/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/first-line-background-images-ignored.tentative.https.sub.html b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/first-line-background-images-ignored.tentative.https.sub.html
new file mode 100644
index 0000000000000000000000000000000000000000..de8186779888830527e5966cbd20f1433e69aeff
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/first-line-background-images-ignored.tentative.https.sub.html
@@ -0,0 +1,44 @@
+<!DOCTYPE HTML>
+<html>
+<head>
+  <title>drawElementImage does not draw cross-origin background images for first-line</title>
+  <script src='/resources/testharness.js'></script>
+  <script src='/resources/testharnessreport.js'></script>
+  <script src='/html/canvas/resources/canvas-tests.js'></script>
+  <script src="/html/canvas/resources/wait-for-canvas-paint.js"></script>
+  <style>
+    div {
+      width: 50px;
+    }
+    *::first-line {
+      color: transparent;
+      background-color: green;
+      background-image: url("https://{{hosts[alt][www]}}:{{ports[https][0]}}/images/red-100x100.png");
+
+    }
+  </style>
+</head>
+<body>
+  <canvas id=canvas width="100" height="200" layoutsubtree>
+    <div id=child>
+      Enough test for several lines of content.
+    </div>
+  </canvas>
+
+  <script>
+    window.onload = () => {
+      promise_test(async function(t) {
+        await waitForCanvasPaint(canvas);
+        const context = canvas.getContext("2d");
+        context.drawElementImage(child, 0, 0);
+
+        // Fetch all pixel data once to avoid multiple slow readbacks.
+        const imgData = context.getImageData(0, 0, canvas.width, canvas.height);
+
+        pixel = _getPixelFromImageData(imgData, 5, 5);
+        assert_array_equals(pixel, [0, 128, 0, 255], "Cross origin should not draw");
+      });
+    };
+  </script>
+</body>
+</html>
diff --git a/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/first-line-background-images-moved-ignored.tentative.https.sub.html b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/first-line-background-images-moved-ignored.tentative.https.sub.html
new file mode 100644
index 0000000000000000000000000000000000000000..527c76dc49d26f21cd4674a7802f9c93fb440240
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/first-line-background-images-moved-ignored.tentative.https.sub.html
@@ -0,0 +1,48 @@
+<!DOCTYPE HTML>
+<html>
+<head>
+  <title>drawElementImage does not draw cross-origin background images for first-line</title>
+  <script src='/resources/testharness.js'></script>
+  <script src='/resources/testharnessreport.js'></script>
+  <script src='/html/canvas/resources/canvas-tests.js'></script>
+  <script src="/html/canvas/resources/wait-for-canvas-paint.js"></script>
+  <style>
+    div {
+      width: 50px;
+    }
+    *::first-line {
+      color: transparent;
+      background-color: green;
+      background-image: url("https://{{hosts[alt][www]}}:{{ports[https][0]}}/images/red-100x100.png");
+
+    }
+  </style>
+</head>
+<body>
+  <canvas id=canvas width="100" height="200" layoutsubtree>
+    <div id=child>
+    </div>
+  </canvas>
+  <div id=grandchild>Enough test for several lines of content.</div>
+
+  <script>
+    window.onload = () => {
+      promise_test(async function(t) {
+        await waitForCanvasPaint(canvas);
+
+        child.moveBefore(grandchild, null);
+
+        await waitForCanvasPaint(canvas);
+        const context = canvas.getContext("2d");
+        context.drawElementImage(child, 0, 0);
+
+        // Fetch all pixel data once to avoid multiple slow readbacks.
+        const imgData = context.getImageData(0, 0, canvas.width, canvas.height);
+
+        pixel = _getPixelFromImageData(imgData, 5, 5);
+        assert_array_equals(pixel, [0, 128, 0, 255], "Cross origin should not draw");
+      });
+    };
+  </script>
+</body>
+</html>
diff --git a/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/first-line-visited-link-color-ignored.tentative.html b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/first-line-visited-link-color-ignored.tentative.html
new file mode 100644
index 0000000000000000000000000000000000000000..0d06e349bd0ee0c23a1dcbc33e1efe56412d833d
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/first-line-visited-link-color-ignored.tentative.html
@@ -0,0 +1,52 @@
+<!DOCTYPE HTML>
+<html>
+<head>
+  <title>drawElementImage does not reveal visited link colors in first-line</title>
+  <link rel="stylesheet" type="text/css" href="/fonts/ahem.css" />
+  <script src='/resources/testharness.js'></script>
+  <script src='/resources/testharnessreport.js'></script>
+  <script src="/html/canvas/resources/wait-for-canvas-paint.js"></script>
+  <style>
+    .link {
+      color: green;
+      display: block;
+      width: 50px;
+    }
+    .link::first-line {
+      background-color: blue;
+      color: blue;
+    }
+    .link:visited {
+      color: red;
+    }
+    .link:visited::first-line {
+      background-color: red;
+      color: red;
+    }
+  </style>
+</head>
+<body>
+  <canvas id="canvas" width="100" height="100" layoutsubtree>
+    <a id=child class=link href="">Enough text for multiple lines.</a>
+  </canvas>
+
+  <script>
+    window.onload = () => {
+      promise_test(async function(t) {
+        await waitForCanvasPaint(canvas);
+
+        var context = canvas.getContext("2d");
+
+        // Draw the element image
+        context.drawElementImage(child, 0, 0);
+
+        // The canvas should NOT reflect the visited state (red), it should paint the unvisited state (green).
+        let pixel = context.getImageData(10, 10, 1, 1).data;
+        assert_false(pixel[0] > 0, "The canvas should not reflect the visited state.");
+        assert_false(pixel[1] > 0, "The canvas should not reflect the visited state.");
+        assert_true(pixel[2] > 0, "The canvas should not reflect the visited state.");
+      });
+    };
+  </script>
+</body>
+</html>
diff --git a/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/first-line-visited-link-color-moved-ignored.tentative.html b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/first-line-visited-link-color-moved-ignored.tentative.html
new file mode 100644
index 0000000000000000000000000000000000000000..6ae77a64d6a6a15b62e8000bd37fd84461d173e4
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/first-line-visited-link-color-moved-ignored.tentative.html
@@ -0,0 +1,58 @@
+<!DOCTYPE HTML>
+<html>
+<head>
+  <title>drawElementImage does not reveal visited link colors in first-line</title>
+  <link rel="stylesheet" type="text/css" href="/fonts/ahem.css" />
+  <script src='/resources/testharness.js'></script>
+  <script src='/resources/testharnessreport.js'></script>
+  <script src="/html/canvas/resources/wait-for-canvas-paint.js"></script>
+  <style>
+    .link {
+      color: green;
+      display: block;
+      width: 50px;
+    }
+    .link::first-line {
+      background-color: blue;
+      color: blue;
+    }
+    .link:visited {
+      color: red;
+    }
+    .link:visited::first-line {
+      background-color: red;
+      color: red;
+    }
+  </style>
+</head>
+<body>
+  <canvas id="canvas" width="100" height="100" layoutsubtree>
+    <div id=child>
+    </div>
+  </canvas>
+  <a id=grandchild class=link href="">Enough text for multiple lines.</a>
+
+  <script>
+    window.onload = () => {
+      promise_test(async function(t) {
+        await waitForCanvasPaint(canvas);
+
+        child.moveBefore(grandchild, null);
+
+        await waitForCanvasPaint(canvas);
+
+        var context = canvas.getContext("2d");
+
+        // Draw the element image
+        context.drawElementImage(child, 0, 0);
+
+        // The canvas should NOT reflect the visited state (red), it should paint the unvisited state (green).
+        let pixel = context.getImageData(10, 10, 1, 1).data;
+        assert_false(pixel[0] > 0, "The canvas should not reflect the visited state.");
+        assert_false(pixel[1] > 0, "The canvas should not reflect the visited state.");
+        assert_true(pixel[2] > 0, "The canvas should not reflect the visited state.");
+      });
+    };
+  </script>
+</body>
+</html>
diff --git a/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/first-line-visited-link-currentcolor-ignored.tentative.html b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/first-line-visited-link-currentcolor-ignored.tentative.html
new file mode 100644
index 0000000000000000000000000000000000000000..56064e09d1fab90eb0306a905c5f4705a47f7281
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/first-line-visited-link-currentcolor-ignored.tentative.html
@@ -0,0 +1,49 @@
+<!DOCTYPE HTML>
+<html>
+<head>
+  <title>drawElementImage does not reveal visited link colors first-line</title>
+  <link rel="stylesheet" type="text/css" href="/fonts/ahem.css" />
+  <script src='/resources/testharness.js'></script>
+  <script src='/resources/testharnessreport.js'></script>
+  <script src="/html/canvas/resources/wait-for-canvas-paint.js"></script>
+  <style>
+    p::first-line {
+      color: currentColor;
+      background-color: currentColor;
+    }
+    .link {
+      font-family: Ahem;
+      font-size: 20px;
+      color: green;
+      display: block;
+      width: 50px;
+    }
+    .link:visited {
+      color: red;
+    }
+  </style>
+</head>
+<body>
+  <canvas id="canvas" width="100" height="100" layoutsubtree>
+    <a id=child class=link href=""><p>Enough text for multiple lines.</p></a>
+  </canvas>
+
+  <script>
+    window.onload = () => {
+      promise_test(async function(t) {
+        await waitForCanvasPaint(canvas);
+
+        var context = canvas.getContext("2d");
+
+        // Draw the element image
+        context.drawElementImage(child, 0, -20);
+
+        // The canvas should NOT reflect the visited state (red), it should paint the unvisited state (green).
+        let pixel = context.getImageData(5, 5, 1, 1).data;
+        assert_false(pixel[0] > 0, "The canvas should not reflect the visited state.");
+        assert_true(pixel[1] > 0, "The canvas should not reflect the visited state.");
+      });
+    };
+  </script>
+</body>
+</html>
diff --git a/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/marker-images-ignored.tentative.https.sub.html b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/marker-images-ignored.tentative.https.sub.html
new file mode 100644
index 0000000000000000000000000000000000000000..9250f67e7d44a83a804ce71f34c94e0658784f13
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/marker-images-ignored.tentative.https.sub.html
@@ -0,0 +1,50 @@
+<!DOCTYPE HTML>
+<html>
+<head>
+  <title>drawElementImage does not draw cross-origin images for ::marker</title>
+  <script src='/resources/testharness.js'></script>
+  <script src='/resources/testharnessreport.js'></script>
+  <script src='/html/canvas/resources/canvas-tests.js'></script>
+  <script src="/html/canvas/resources/wait-for-canvas-paint.js"></script>
+  <style>
+    canvas {
+      background-color: black;
+    }
+    ::marker {
+      content: url("https://{{hosts[alt][www]}}:{{ports[https][0]}}/images/red-16x16.png");
+    }
+  </style>
+</head>
+<body>
+  <canvas id=canvas width="100" height="100" layoutsubtree>
+    <ul id=child>
+      <li>One</li>
+      <li>Two</li>
+    </ul>
+  </canvas>
+
+  <script>
+    window.onload = () => {
+      promise_test(async function(t) {
+        await waitForCanvasPaint(canvas);
+        const context = canvas.getContext("2d");
+        context.drawElementImage(child, 0, 0);
+
+        // Fetch all pixel data once to avoid multiple slow readbacks.
+        const imgData = context.getImageData(0, 0, canvas.width, canvas.height);
+
+        // Platforms differ on where them arker is drawn. So ensure there is no red any place.
+        for (y = 0; y < 80; y++) {
+          y_offset = y * canvas.width;
+          for (x = 0; x < 40; x++) {
+            x_offset = (y_offset + x) * 4;
+            for (channel = 0; channel < 4; channel++) {
+              assert_false(imgData.data[x_offset + channel] > 0, "Cross origin ::marker should not draw");
+            }
+          }
+        }
+      });
+    };
+  </script>
+</body>
+</html>
diff --git a/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/marker-visited-link-currentcolor-ignored.tentative.html b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/marker-visited-link-currentcolor-ignored.tentative.html
new file mode 100644
index 0000000000000000000000000000000000000000..9c3be2e984c60f1d7229f632a95e18b91d7659c9
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/html/canvas/element/manual/draw-element-image/privacy/pseudos/marker-visited-link-currentcolor-ignored.tentative.html
@@ -0,0 +1,51 @@
+<!DOCTYPE HTML>
+<html>
+<head>
+  <title>drawElementImage does not reveal visited link colors in ::marker</title>
+  <script src='/resources/testharness.js'></script>
+  <script src='/resources/testharnessreport.js'></script>
+  <script src='/html/canvas/resources/canvas-tests.js'></script>
+  <script src="/html/canvas/resources/wait-for-canvas-paint.js"></script>
+  <style>
+    ::marker {
+      content: "\25AE";
+      color: currentColor;
+      font-size: 50px;
+    }
+    .link {
+      color: green;
+    }
+    .link:visited {
+      color: red;
+    }
+  </style>
+</head>
+<body>
+  <canvas id="canvas" width="100" height="100" layoutsubtree>
+    <a id=child class=link href="">
+      <ul>
+        <li>One</li>
+        <li>Two</li>
+      </ul>
+    </a>
+  </canvas>
+
+  <script>
+    window.onload = () => {
+      promise_test(async function(t) {
+        await waitForCanvasPaint(canvas);
+
+        var context = canvas.getContext("2d");
+
+        // Draw the element image
+        context.drawElementImage(child, 0, 0);
+
+        // The canvas should NOT reflect the visited state (red), it should paint the unvisited state (green).
+        let pixel = context.getImageData(20, 40, 1, 1).data;
+        assert_false(pixel[0] > 0, "The canvas should not reflect the visited state.");
+        assert_true(pixel[1] > 0, "The canvas should not reflect the visited state.");
+      });
+    };
+  </script>
+</body>
+</html>
diff --git a/third_party/blink/web_tests/wpt_internal/html/canvas/drawElementImage/privacy/resources/subframe-cross-origin-before-background-on-body-modified.https.sub.html b/third_party/blink/web_tests/wpt_internal/html/canvas/drawElementImage/privacy/resources/subframe-cross-origin-before-background-on-body-modified.https.sub.html
new file mode 100644
index 0000000000000000000000000000000000000000..11ab63c029fb2eb2c01216146ce898161181b377
--- /dev/null
+++ b/third_party/blink/web_tests/wpt_internal/html/canvas/drawElementImage/privacy/resources/subframe-cross-origin-before-background-on-body-modified.https.sub.html
@@ -0,0 +1,22 @@
+<!DOCTYPE HTML>
+<html>
+<head>
+  <title>subframe with cross-origin image in ::before on the body</title>
+  <style>
+    body.has-before::before {
+      content: "";
+      width: 100px;
+      height: 100px;
+      background-image: url("https://{{hosts[alt][www]}}:{{ports[https][0]}}/images/red-100x100.png");
+      background-color: green;
+      display: block;
+    }
+    div {
+      width: 100px;
+      height: 100px;
+      background: transparent;
+    }
+  </style>
+</head>
+<body><div></div></body>
+</html>
diff --git a/third_party/blink/web_tests/wpt_internal/html/canvas/drawElementImage/privacy/resources/subframe-cross-origin-before-background-on-body.https.sub.html b/third_party/blink/web_tests/wpt_internal/html/canvas/drawElementImage/privacy/resources/subframe-cross-origin-before-background-on-body.https.sub.html
new file mode 100644
index 0000000000000000000000000000000000000000..0ecc58cfbe8d4481527a81bb3c6b2e766561fb21
--- /dev/null
+++ b/third_party/blink/web_tests/wpt_internal/html/canvas/drawElementImage/privacy/resources/subframe-cross-origin-before-background-on-body.https.sub.html
@@ -0,0 +1,22 @@
+<!DOCTYPE HTML>
+<html>
+<head>
+  <title>subframe with cross-origin image in ::before on the body</title>
+  <style>
+    body::before {
+      content: "";
+      width: 100px;
+      height: 100px;
+      background-image: url("https://{{hosts[alt][www]}}:{{ports[https][0]}}/images/red-100x100.png");
+      background-color: green;
+      display: block;
+    }
+    div {
+      width: 100px;
+      height: 100px;
+      background: transparent;
+    }
+  </style>
+</head>
+<body><div></div></body>
+</html>
diff --git a/third_party/blink/web_tests/wpt_internal/html/canvas/drawElementImage/privacy/resources/subframe-cross-origin-before-content-on-body.https.sub.html b/third_party/blink/web_tests/wpt_internal/html/canvas/drawElementImage/privacy/resources/subframe-cross-origin-before-content-on-body.https.sub.html
new file mode 100644
index 0000000000000000000000000000000000000000..88b7311f8626d1f8d38662cc1dd3e90701d960f5
--- /dev/null
+++ b/third_party/blink/web_tests/wpt_internal/html/canvas/drawElementImage/privacy/resources/subframe-cross-origin-before-content-on-body.https.sub.html
@@ -0,0 +1,23 @@
+<!DOCTYPE HTML>
+<html>
+<head>
+  <title>subframe with cross-origin image in ::before on the body</title>
+  <style>
+    body::before {
+      content: url("https://{{hosts[alt][www]}}:{{ports[https][0]}}/images/red-100x100.png");
+      width: 100px;
+      height: 100px;
+      background-color: green;
+      display: block;
+    }
+    div {
+      width: 100px;
+      height: 100px;
+      background: transparent;
+    }
+  </style>
+</head>
+<body>
+  <div></div>
+</body>
+</html>
